File: LeetCode/Python/0022. Generate Parentheses.py
class Solution:
    def generateParenthesis(self, n: int) -> List[str]:
        all_combinations = []
        self._generateParenthesis(n, 0, 0, "", all_combinations)
        return all_combinations
    
    def _generateParenthesis(self, n, opened, closed, s, all_combinations):
        if opened > n:
            return
        if closed > opened:
            return
        if opened == closed == n:
            all_combinations.append(s)
        else:
            self._generateParenthesis(n, opened + 1, closed, s + "(", all_combinations)
            self._generateParenthesis(n, opened, closed + 1, s + ")", all_combinations)

File: LeetCode/Java/0022. Generate Parentheses.java
class Solution {
    public List<String> generateParenthesis(int n) {
        ArrayList<String> allCombinations = new ArrayList<>();
        generateParenthesis(n, 0, 0, "", allCombinations);
        return allCombinations;
    }
    
    private void generateParenthesis(int n, int opened, int closed, String s, ArrayList<String> allCombinations) {
        if (opened > n)
            return;
        if (closed > opened)
            return;
        if (opened == n && closed == n) {
            allCombinations.add(s);
            return;
        }
        
        generateParenthesis(n, opened + 1, closed, s + "(", allCombinations);
        generateParenthesis(n, opened, closed + 1, s + ")", allCombinations);
    }
}
